src/rgw/rgw_string.h

```diff
@@ -233,4 +238,167 @@ extern bool match_wildcards(boost::string_view pattern,
                             boost::string_view input,
                             uint32_t flags = 0);
 
+inline void rgw_escape_str(const std::string& s, const char esc_char, const char special_char, std::string& dest)
+{
+ dest.reserve(s.size() * 2);
+ 
+ auto dest_iter = begin(dest);
+ 
+ for (const char c : s) {
+    if(esc_char == c || special_char == c) {
+        *dest_iter++ = esc_char;
+    }
+    
+    *dest_iter++ = c;
+ }
+}
+
+inline void rgw_escape_str(const std::string& s, const char esc_char, const char special_char, std::string *dest_ptr)
+{
+ return rgw_escape_str(s, esc_char, special_char, *dest_ptr);
+}
+
+inline std::string rgw_escape_str(const std::string& s, const char esc_char, const char special_char)
+{
+ std::string result;
+ rgw_escape_str(s, esc_char, special_char, result);
+ return result;
+}
+
+inline ssize_t rgw_unescape_str(const std::string& src, const ssize_t ofs,
+                       const char esc_char, const char special_char,
+                       std::string& dest)
+{
+  dest.reserve(src.size());
+  auto destp = begin(dest);
+
+  bool esc = false;
+
+  for (size_t i = ofs; i < src.size(); i++) {
+    char c = src[i];
+    if (!esc && c == esc_char) {
+      esc = true;
+      continue;
+    }
+    if (!esc && c == special_char) {
+      dest.resize(i);
+      return dest.size();
+    }
+    *destp++ = c;
+    esc = false;
+  }
+
+  return std::string::npos;
+}
+
+inline ssize_t rgw_unescape_str(const std::string& s, const ssize_t ofs,
+                       const char esc_char, const char special_char,
+                       std::string *dest_ptr)
+{
+  return rgw_unescape_str(s, ofs, esc_char, special_char, *dest_ptr);
+}
+
+class CephContext;
+
+/* size should be the required string size + 1 */
+int gen_rand_base64(CephContext *cct, char *dest, int size);
+
+void gen_rand_alphanumeric_lower_CHAR(CephContext *cct, char *dest, const int nchars);
+
+namespace detail {
+
+// Does NOT null-terminate:
+template <typename OutputIterT, typename SizeT>
+void gen_rand_buffer_mutate_in_place(OutputIterT& oi, const SizeT size, std::string_view tbl)
+{
+  std::generate_n(oi, size, [&]() noexcept {
+                  auto n = ceph::util::generate_random_number(0, tbl.size());
```

Comment:
Aaaah, I see what's going on there-- thank you for point that out! I'll rework this. I was /hoping/ to be able to keep the CephContext* out of these (and, maybe that's still possible for most cases), let me tinker with it. The original call to random()->get_bytes() is in CryptoRandom, and it looks like it either calls getentropy() or reads from /dev/urandom.

Reply:
Question: 
Would there be another standard random engine that would be equally satisfying? For example, std::random_device? (http://www.cplusplus.com/reference/random/random_device/) I note that getentropy() can also fail (and indeed, in libc++ at least I see basically the same implementation behavior as we're using-- random_device tries to call getentropy() or use /dev/urandom). It's easy to get the Ceph random library to use another engine.

Reply:
there's some more background and benchmarks in https://github.com/ceph/ceph/pull/17972, which was motivated by http://tracker.ceph.com/issues/21401. you're welcome to explore other options there

Reply:
Ok, thank you! I'll explore. It would be easier to consider the two situations separate. However, one guide might be in asking whether or not it is /desirable/ to use the standard library here, if possible. 
Basically include/random may or may not be sufficient to handle this (I think it is, if we "merely" need to change the engine), or maybe not if we're doing something truly specialized, but I'm thinking we're not. I'll keep in touch with you on this.

Reply:
the standard library is desirable if it meets the same criteria of multithreaded performance, memory usage, and use of file descriptors

i'm wary of using thread-local storage for heavyweight engines like std::mt19937 (5000 bytes in size), because radosgw's thread pool size currently defaults to 512

another consideration is that these functions generate a _stream_ of bytes - getentropy() and /dev/urandom both do this in bulk, rather than having to generate a separate index for each byte

Reply:
@cbodley Opinion here? I have a slight preference to make this another PR (frankly just because I need to move on to some other things for now), but can also implement g_ceph_context-- I need to do some hunting around to see what that might affect.

Reply:
it makes sense for those helper functions to take CryptoRandom instead of CephContext, but i wouldn't worry it past that